"Any" invoice

engenegr opened the pull request:
Hey. I needed to generate 0-amount invoices inside our system and I know that Core Lightning generally allows that. The only problem was that it doesn't allow "0" amount but demands "any" string in `invoice`.

I am new in Rust and not sure that my solution is ideal. I am proposing this change only because it may be convenient to stick with `cargo` repository distributed version of the library.

vincenzopalazzo reviewed (changes requested):
Hi, 

Sorry for the delay here!

>I am proposing this change only because it may be convenient to stick with cargo repository distributed version of the library.

Yes please, keep this change in a fork do not help.

I have some small requests in the API, but also a test could be good to add. This can be implemented like this

```diff
diff --git a/rpc/tests/rpc_utility_test.rs b/rpc/tests/rpc_utility_test.rs
index af9d5cf..10b8bb4 100644
--- a/rpc/tests/rpc_utility_test.rs
+++ b/rpc/tests/rpc_utility_test.rs
@@ -90,3 +90,10 @@ fn listinvoice_by_payment_hash_test_one(lightningd: LightningRPC) {
     let listinvoice = lightningd.listinvoices(None, None, None, None);
     assert!(listinvoice.unwrap().invoices.is_empty());
 }
+
+#[rstest]
+fn generate_any_invoice_test_one(lightningd: LightningRPC) {
+    let invoice = lightningd.invoice(None, "anyinvoice", "generate an any invoice", None).unwrap();
+    let decode = lightningd.decodepay(invoice.bolt11, None).unwrap();
+    assert_eq!(decode.amount_msat, None);
+}
```

I did not test! maybe running in the ci helps.

As last request could you please add in the commit header the following message `fix(rpc): be able to generate invoice with any amount`.

P.S: use the rpc typed I think is super painful because I'm not maintaining it because I'm working in a code generation to generate the models. I suggest to you to use the [common crate](https://github.com/laanwj/cln4rust/tree/master/common) that allow to write your struct and encode if directly without relay in unstable and not completed API  in rpc. This is a good example https://github.com/vincenzopalazzo/cln-remote-api/blob/main/rest-api/src/info.rs

This is an example of a generics method with common crate https://github.com/vincenzopalazzo/cln-remote-api/blob/main/rest-api/src/utils.rs

engenegr commented on rpc/src/lightningrpc.rs:
```diff
@@ -144,15 +144,26 @@ impl LightningRPC {
-                expiry,
-            },
-        )
+        match msatoshi {
+            0 => self.call(
+                "invoice",
+                requests::AnyInvoice {
+                    msatoshi: "any",
+                    label,
+                    description,
+                    expiry,
+                },
+            ),
+            _ => self.call(
+                "invoice",
+                requests::Invoice {
+                    msatoshi,
+                    label,
+                    description,
+                    expiry,
+                },
+            ),
+        }
```
Yes. But isn't that 0 u64 will match as Some(0)?

engenegr replied:
Nevermind though. I will add test and see what happens.

vincenzopalazzo replied:
With my solution, I was going to reflect the behavior of cln

```
➜  ~ clightning invoice 0 "zeroinvoice" "zero invoice"               
{
   "code": -32602,
   "message": "amount_msat|msatoshi: should be positive msat or 'any': invalid token '0'"
}
```

IMHO 0 as amount needs to be an error, and None be converted to any

engenegr replied:
So you are in Core Lightning camp.
I don't like number-any solution because it makes API weird. I will look at the code again and return to this PR.

vincenzopalazzo commented on rpc/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 #clightningrpc-common = { path = "../common" }
 clightningrpc-common = "0.3.0-beta.2"
+uuid = { version = "0.8.2", features = ["v4"]}
 
 [dev-dependencies]
 rstest = "0.10.0"
```
```suggestion

[dev-dependencies]
rstest = "0.10.0"
uuid = { version = "0.8.2", features = ["v4"]}
```

engenegr replied:
Ah, sure.

engenegr commented:
@vincenzopalazzo 

I needed to add something for tests since I've struggled with testing it without bootstrapping additional environment I don't need. I think I could contribute more in that regard but later in another PR.

I've added random invoice labels and changed the code as you suggested.

vincenzopalazzo reviewed (changes requested):
Minor change and I think we are ready to go

vincenzopalazzo reviewed (approved):
ACK https://github.com/laanwj/cln4rust/pull/63/commits/cd5845b7c7e7099f7f14f3ab0cc286d68b1c9c72

The CI  failure is unrelated!

Thanks!
